### ISSUE TITLE
Pthreads/parallel both useless

### ISSUE THREAD

**Issue by joeyhub** [joeyhub/aphpsync#2]
After experimenting with pthreads, the results of my experiments have been grim, that the library simply if not in stable nor maintainable state to be put to task.

The maintainer and contributors have also ceased support for pthreads to be replaced with another implementation. In many ways the new effort is a step in the right direction yet unfortunately stunted such that it wont be viable for what I want to achieve.

I was a fun experiment but I also lack confidence in the maintainers of those libraries to deliver a complete MVP for multiple threading in PHP that's suitable for general use anytime soon. This is compounded by that the maintainers are limited in their capacity to receive criticism and feedback.

It's a bit strange to me to see that over the course of five years on and off huge efforts put in to give PHP threading by people that would presumably be using threads and heavily into threads yet in all that time it appears certain crucial basics required to cover all of the major bases have not been picked up along the way.

To under stand the problem...

Code:

    $threads = [new Thread(), new Thread()];
    $futures = [$threads[0]->run(function(){sleep(4);return null;}), $threads[2]->run(function(){sleep(0);return null;})];
    foreach($futures as $future) {
        echo "{$future->value()}\n";
        sleep(4);
    }

Expected: 8 seconds

Actual: 12 seconds

The implication of this that rather than achieving a speed of multiple threads, you achieve a speed of anywhere between one thread to how many threads you have depending on circumstances. If you have a solution giving multiple threading but it gives you single threaded performance then that's an issue.

The main thread here ends up being delayed by as much time as the first thread took to run. You have situations where you can process ten workloads which can all end up being as slow as the slowest thread. That means that you might have nine that would take a second but one that takes an hour. Because of that one taking an hour now they all take an hour.

The reason for this is that it blocks per thread. If you want to implement multi-threading, per thread might raise alarms as there's a good chance that's not multi-threaded. The simplest common solution to this which delivers close to the maximum achievable efficiency is to use multiplexing and demultiplexing. More simply put your blocking function should work with multiple threads. Otherwise it's not fully multi-threaded.

Traditionally that might look like:

```while(($future = Futures::select($futures, $timeout)) !== null)```

Rather than the futures being iterated in a pre-determined order, the select function yields each future in the order in which they actually become available.

There are a few keep reasons for using multiple threads. That includes working around problems with blocking calls, minimising latency and improving performance especially by not having CPU wasted when it could be working on IE. There's almost no functional reason to use threads. It's nearly all about performance. The only exception might be where you have calls that can block forever leading to hanging.

This library has a blocking call, that can block forever and that can only operate on one resource at a time, exactly the same as the blocking calls we would want to use threads to avoid!

In this sense it's barely fit for purpose. Unless you either don't care to receive the result, no know if or when more than one child thread finished or the thing that's waiting for them depends on both of them then the library can't be used. It's only applicable for a subset of parallel processing scenarios.

If you want to ask why use php threads instead of processes, that answer to that is that it's like multiple processes but with less overhead. However because of the problem seen about, it works out to be significantly faster and more efficient to use multiple processes instead which for now is what I'll do instead.

**Comment by joeyhub** [comment]
The solution to this is, go back to the basics and use proc instead. I can keep overhead down potentially if I have some basic global process pool management for the horizontal threading. That is to be able to say one process for mysql, one for curl, or to be able to say two for mysql. Sort of gets a bit complex though sort of load balancing which can't be done perfectly as things must be sticky. The worse case would probably be ending up with a matrix of a * b threads.

**Comment by joeyhub** [comment]
The maintainers have started to experiment now with incorporating select in the parallel extention design so there's some hope yet for recovery of that effort into something decent enough.

Though considering what would be ideal for async...

There's some question whether it's better to use simply a top down unidirectional callback based mechanism or inter thread communication. Function stacks are simpler as well as trees. I think I might have answered that design query already with the one thread dedicated to a given purpose scenario that might want to be accessible anywhere at any point a thread needs that.

Irksome considerations...

Traditionally multi-process PHP also is one to one with the PHP engine context. One process, one context. When we split that, where do you draw the line?

There are somethings that might want to be kept in common or duplicated or even wrapped (mutexted as shared, etc).

Many of those complexities are hard to truly know without knowing fully the internals.

People are thinking about putting in a bootstrap. For me that's a big unknown. Why can't you put bootstrap in run as the first task?

For me that's potentially an area of concern.

* One Process
* Many Contexts
* Start Context
* One thread per context
* Thread must have context to run

When a PHP Context dies should that be considered non-recoverable?
When a thread dies that's not really so?
Though a thread death can also kill a context.
It makes no sense to me that a context can die other than during the execution of its thread.

It's very tricky whether to say an exception in a thread breaks the context or not.
Either way we need an understanding of how execution ended.
Normally this isn't a problem, for a process we'd get a bad exit code.
For a PHP context 

Caveat might be creation. How much overhead is involved in creating a context?

If you have thread inter communications (unidirectional pipes lets say), do you critically need to return?

Memory management for threads? If they can all reference each other then which is really dominant?

interface Selectable;
interface Stateful
{
	const STATE_*;
	abstract const StateSet STATES;
}
interface StatefulSelectable extends Selectable, Stateful;

class Select implements StatefulSelectable
{
	// This is a joke.
	const STATES = [IDLE, BUSY];

	// Selects assoc of STATE => Selectable[]
	state($timeout, $selects): [STATE, Selectable][]
}

No from? No edge type?

class PhpContext implements Selectable
{
	const STATES = [CREATING, IDLE, BUSY, DESTROYING, FUBAR, DESTROYED];

	public function __construct(Config $processContext, Config $phpContext);

	// !DESTROY* !FUBAR
	add(Closure $main): Thread
	// BUSY
	purge():void

	// Same as __destruct?
	// IDLE BUSY?
	destroy()
	// IDLE
	__clone()
}

CREATING [from start]
CREATING -> IDLE [successfully created]
IDLE -> BUSY [when empty and when closure added ]
BUSY -> IDLE [when thread finished and when empty]
IDLE -> DESTROYING [when DESTROY called or potentially destruct]
BUSY -> DESTROYING [when catastrophy during BUSY]
CREATING -> DESTROYING [when catastrophy during creating]
DESTROYING -> DESTROYED [when destroying is finished]
DESTROYING -> FUBAR [when catastrophy during destroying]
BUSY -> BUSY [when next thread]

Great thing with this, B&B.

Raises a fun question, an error during destroying really is catastrophic.

Also raises the question of do we want a queue object and some queue management or is just remove enough?
We probably really want the ability to just wipe it all out and empty it.

clone might be a great performance boost but also playing with fire needing stuff like updating pointers after memcpy.
also needs to lock == block? But what if PHP has it build in?

Stuff like that, thread management/pool can be custom managed in a dedicated thread?

class Thread implements Selectable
{
	const STATES = [PENDING, FINISHING, DYING, FINISHED, DIED, REMOVED];

	// PENDING
	remove():void;
	// RUNNING
	kill():void;

	// FINISHED DIED REMOVED (exceptions for last two)
	result():mixed!Exception
}

PENDING [from start]
PENDING -> FINISHING [Links to PhpContext * to BUSY]
FINISHING -> DYING [when catastrophy or kill called, needs FUBAR as well?]
FINISHING -> FINISHED // after FINISHING or even DYING if exceptions in use?
DYING -> DIED [From DYING or catastrophy during finishing]
PENDING -> REMOVED [when remove called or empty on PhpContext]

Class structure, many php contexts, each has one queue, each queue has zero or more jobs, php context has zero or one running job, thread concerned with queue parent and context parent

States can all be grouped? into head, body, tail or begin, middle, end. Bubbling is probably a bit much except those groups. if groups states can be selected for or even any state change...
wants getState on stateful? Can also overlay that with canX.

Issue: States can change while handling states.

What states are actually useful? Some can be made in userland as in the user knows they call something changing the state immediately.

call_user_func(function() {
	Thread::loop();
});

There's a battle for dominance here. As I view it, the extension context should be dominant. One context leads to another so things end when all contexts end and don't lead to another. GC of thread or context in a scope should not gc the fascade object representing it in that running php context. The extension scope wants to gc contexts when their state is that they're GCed from all active contexts and are inert? When they become inert but still reffed then they should turn refs into mere fascades no longer connecting to extension scope, just erroring on any call that would do that due to wrong state, transferring to local. It would be on the user to call destroy if they're really done with it.

The battle is that the initial context isn't necessarily made by the extension and something else might be clinging to it as well as awaiting for a result from it. We don't want to waste that original context but there needs to be some rules or mechanism of control. Really in that case you want when the main context to finish for it to be the last.

Should there be a static class/functions for the extension context? The main context code could also be used in the contexts it creates?

Some points if using closures...

$c = function() {};foreach(something)$context->run($c);

most probably beats

foreach(something)$context->run(function() {});

Passing data and pipes... might be easier to just say no use, data passed through rest of params then are param to closure.

What about ability to compile/parse once rather than maybe implicit caching?

Such as Job::from(Closure) that would validate once and only once. then ->run($job->pass($data)) same as ->run($job, $data);?

Data passing should basically be same as serialization but under the hood it can be direct single pass and memcpy or something instead.

Think about take igbinary, copy the serialize bit, but then replace each bit outputing the byte with the bit taking the action from the byte from the unserialize.

Same for pipes which is fun. Something simple might be to have a funnel per thread. So that thread has ->messages() would be blocking and return array of messages.
Then message which would push a message, IE it's just a queue, mutexed probably. I don't see why to send a message can't be called from anywhere. Single consumer many producers.
That's a funnel, simplex, demuxed. That can actually be used by multiple threads, first come first serve. They might want to shift off instead or take a proportion (IE based on listeners but really should be explicit)? You then might want the reverse, to broadcast or worse many to many. Full duplex is just two simplex. If you have the many to many technically you can support everything?  I think always many to [many, single], though single can actually be to many, it's just deduped though/serialized.

This really makes me thing pipes want to be separate objects where listeners can be added and removed. Anyone with a ref is a potential producer? Each broadcast listener ends up needing their own queue. Had this with just using stream select. There was at one point a question of should an attachment be for all time or attach time with the latter usually prevailing.

**Comment by joeyhub** [comment]
That's speculative, for the proc approach there was no problem with pipes though everything was done inside a single master process or context so no P2P facilitated through that (though was possible to technically speaking give a process another's pipe), instead it was a hub or middle man for everything which might be an interesting consideration for both proc and threads.

For either approach, that's a bit of a mix of potential there and overlapping concerns. There's no the question here though of when there are other root processes delegated to handled and how to find them or their communication points.

For the proc approach pipes is really best quites for parent/child relations. There's a question of what other mechanisms might be relevant, perhaps regardless of whether or not pipes are also used or whether the relationship is parent and child. There's obviously network, other intermediate messaging services and unix sockets but what really works best are things probably globally findable and with very little overhead. It would be interesting to consider how some of those IPC mechanisms might play with single process and multiple contexts.

There's quite a lot around with IPC and my approach there would probably be, get lost windows and other operating systems that are just too weird and unlikely to worry about (linux is free so rare valid excuse for needing windows).

eio, sync, semaphore and shared memory are all worth some experimentation at some point.

pht is something I completely missed which might be an exciting potential alternative to parallel/pthreads. I think the maintainer for that is also more stable (though may lack time to fully maintain it).

**Comment by joeyhub** [comment]
Bullying.

https://chat.stackoverflow.com/transcript/11/2019/2/17/0-20

I originally tried pthreads with it's quality falling far lower than I expected even for PHP.

I spoke out about that, honestly. I thought the project had basically been abandoned to be honest and was open game.

How do you know that they're all the villains? Not a single one could come and face me directly which is how this all started, me facing the author of a library directly that I found lacklustre (I liked their response to restart it but outside of that this specimen was difficult to say the least).

Then the author suddenly turned up and said scrap this, then proposed a new design. The new design was in fact excellent solving almost all of the problems from the original. I think one of the major problems from the original was incorporating concepts from elsewhere rather than presenting the concept as it is to PHP. For example want to be threaded objects like in Java and this volatile stuff from elsewhere that just doesn't really mix, it's setting yourself up.

However the new version had a missing piece so I went and tried to explain that and promote the case for it. It was a major turnaround barring this one massive oversight. The moment I did the author started getting ultra defiant and not appearing to understand so I tried very hard. Then they lost it and deleted those threads so you can't see. To be fair, I reached a point of frustration that I went quite far with the persuasion but not initially. I also had other things going on and wasn't entirely a happy bunny but that's not how things started.

Looking back at that, I should not entirely have assumed it was as obvious to others as it was to me. I'm a retard. I've just found out that the geniuses as MS and Google have missed a few points on this very basic feature I want. What the fuck? I belong in a gulag. I'm not even politically correct. Fuck. I'm a criminal. I think to myself. That's worse than talking to yourself.

After deleting all the initial comments this person then having lost the argument goes to their mates to indirectly yet publicly talk shit about me. Why? I can do that myself. This is a single character for all their flaws and they're trying to do well I mean Christ it's not like the whole of MS with their however many billion in the pocket failing sending a proposal to TC39 that passed despite being fundamentally flawed.

Most of what this character has to say are lies but some of it is more subtle. They say often I'm wrong and confused. Yes I am and I call myself out first. It simply echoed my own statements (such as declaring my own brainfart).

What I really don't appreciate is the lies. I posted this specifically after a comment was marked as spam which wasn't. That's outright lying. That's dishonesty. Mark it as too heated then we don't have a problem but lie and yes, it becomes personally, if you impose a negative aspect of your personality on others, abusing privileges then you will be judged and that judgement will be shared. Though I can't entirely fault the individual now seeing that this has become a trend. Part of the problem is the ivory toward except it's not ivory it's the fan boys and except they're not defending you from bullies but from people they've bullied that are then retaliating.

That aside, regardless of the emotional turmoil this individual has managed to focus their efforts towards progressively better results and that is commendable. Regardless of their flimming and flamming they're working on something that could help PHP and be a game changer. However, expectations are also high and they need to learn to deal with that. It's also a means to an end.

This cunt doesn't deserve recognition for their social displays but their technical contributions to humanity could take them far.

There are other areas of consideration they could work on and eventually achieve great success. One of those is looking at all the async libs in PHP and unifying them or even presenting an analysis from under the hood.

Currently in PHP to have proper async many implementations (unless I've missed some) basically just replace native C based implementations with native PHP based implementations ultimately to only be able to get a resource to stick into a multiplexor like select (FS into epoll, etc) or stream_select.
